### PR TITLE
Add DynamoDB and Kinesis Time Window events and response models

### PR DESCRIPTION
**Pull request by carlzogh**
*Issue #, if available:*
https://github.com/aws/aws-lambda-go/issues/358

*Description of changes:*
Add support for Time Window events.

Based on Java implementation: https://github.com/aws/aws-lambda-java-libs/pull/202

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by zsombor-balogh — APPROVED** [review]

**Review by smirnoal — DISMISSED** [review]

**Review by andclt — DISMISSED** [review]

File: events/time_window.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
+package events
+
+// Window is the object that captures the time window for the records in the event when using the tumbling windows feature
+// Kinesis: https://docs.aws.amazon.com/lambda/latest/dg/with-kinesis.html#services-kinesis-windows
+// DDB: https://docs.aws.amazon.com/lambda/latest/dg/with-ddb.html#services-ddb-windows
+type Window struct {
+	Start string `json:"start"`
+	End   string `json:"end"`
```
[the remark this answers is not in the document]
Comment (carlzogh) [reply events/time_window.go]: Thanks for this comment, I've pushed new commits with a string -> time.Time ser/deser (uses `RFC3339` as defined in https://cs.opensource.google/go/go/+/refs/tags/go1.18.1:src/time/format.go;l=101)

File: events/time_window.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	Window Window `json:"window"`
+
+	// State being built up to this invoke in the time window.
+	State map[string]string `json:"state"`
+
+	// Shard id of the records
+	ShardID string `json:"shardId"`
+
+	// Dynamodb stream arn.
+	EventSourceARN string `json:"eventSourceARN"`
```
[the remark this answers is not in the document]
Comment (carlzogh) [reply events/time_window.go]: Good spotting, fixed

File: events/time_window.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+type KinesisTimeWindowEventResponse struct {
+	TimeWindowEventResponseProperties
+	BatchItemFailures []KinesisBatchItemFailure `json:"batchItemFailures"`
+}
+
+// DynamoDBTimeWindowEventResponse is the outer structure to report batch item failures for DynamoDBTimeWindowEvent.
+type DynamoDBTimeWindowEventResponse struct {
+	TimeWindowEventResponseProperties
+	BatchItemFailures []DynamoDBBatchItemFailure `json:"batchItemFailures"`
+}
```
[the remark this answers is not in the document]
Comment (carlzogh) [reply events/time_window.go]: I didn't really intend for those to end up there, but I guess I must have forgotten to move them (creating them was a little bit of a copy-paste activity in this file)

File: events/epoch_time.go [hunk 4]
```diff
@@ -57,3 +62,24 @@ func (e *MilliSecondsEpochTime) UnmarshalJSON(b []byte) error {
...
+
+func (e *RFC3339EpochTime) UnmarshalJSON(b []byte) error {
+	var isoTimestampStr string
+	err := json.Unmarshal(b, &isoTimestampStr)
+	if err != nil {
+		return err
+	}
+
+	parsed, _ := time.Parse(time.RFC3339, isoTimestampStr)
+	if err != nil {
```
Comment (bmoffatt) [inline events/epoch_time.go:79]: lost the error!

Comment (carlzogh) [reply events/epoch_time.go]: 🤦🏻found it again in the next commit!

**Review by bmoffatt — APPROVED** [review]

**Pull request merged**